### DISCUSSION TITLE
Data type problem in subject_id and causer_id

### DISCUSSION THREAD

**Discussion by yangshengjie4u** [spatie/laravel-activitylog#932] [Q&A] [answered]
I'm going to change the primary key to uuid for some tables. It can cause the mixed data type (int/string) for `subject_id` and `causer_id` columns in the `activity_log` table. 
Any ideas to resolve this issue? Should I keep all primary key as same data type? If I change all primary key to uuid, how to change the `subject_id` and `causer_id` fields to string?

**Comment by Gummibeer** [comment, marked as answer]
Hey,
yes it makes things A LOT easier if all your PKs have the same type. You can change the `causer` and `subject` relationship columns in the published package migration. 
https://github.com/spatie/laravel-activitylog/blob/0da6f05060364bb9a15c528d459445d52606d781/database/migrations/CreateActivityLogTable.php.stub#L15-L16

Instead of `nullableMorph()` there should also be a `nullableUuidMorph()` or something like that.

**Reply by colinmackinlay** [reply]
@Gummibeer do you know if  it is possible to have both types of PK for subject? I log activity on multiple models whcih have integer ID's but do have one model which has a UUID which I want to log activity for too.

**Reply by Gummibeer** [reply]
Depends on your database and how strict it is with types. You could create simple string columns instead of big int or UUID.
But it will decrease performance. And it's not recommended - you should use one PK system in your app.

**Reply by colinmackinlay** [reply]
Thanks - thought so. Integers are easiest but UUID would be better for some records that contain personal data. I can see why it would be an issue though

**Reply by Gummibeer** [reply]
You don't have to use UUIDs as PKs only for routing to prevent numbering attacks.
You can have a secondary key as UUID and use that one for routing. That way you can use the big int AI as PK everywhere and use the UUID for routing.
But keep in mind: the UUID isn't any kind of security as that should be in your policies. In theory I could even guess UUIDs, primarily ordered ones, and do a brute-force attack. So the UUID in URL is only a cosmetical thing and doesn't really contribute to security.

If you want to use integers and mask them in URL you could also check [Hash-IDs](https://sqids.org/php) (the ones like YouTube video IDs). Or checkout Twitter [Snowflake-IDs](https://en.wikipedia.org/wiki/Snowflake_ID).
